In [1]:
import keras
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split as tts

from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [29]:
concrete = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concreteX = concrete[concrete.columns[0:-1]]
concreteY = concrete[concrete.columns[-1]]
X_train, X_test, y_train, y_split = tts(concreteX, concreteY, test_size=0.3)


In [26]:
from sklearn.metrics import mean_squared_error
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(concreteX.shape[1],)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

## First Question 50-MSE-STDEV

In [56]:
mselist = []
for count in np.arange(0,50):
    X_train, X_test, y_train, y_split = tts(concreteX, concreteY, test_size=0.3)
    model = regression_model()
    model.fit(X_train, y_train, epochs=50, verbose=0)
    yhat = model.predict(X_test)
    mse = mean_squared_error(yhat,y_split)
    mselist.append(mse)

In [58]:
from statistics import mean, stdev
print('_'*50)
print('The mean of the MSE is {}'.format(mean(mselist)))
print('_'*50)
print('The Standard Deviation is {} '.format(stdev(mselist)))
print('_'*50)


__________________________________________________
The mean of the MSE is 379.60411795302457
__________________________________________________
The Standard Deviation is 551.0850260383576 
__________________________________________________


## Answer

__________________________________________________
The mean of the MSE is 379.60411795302457
__________________________________________________
The Standard Deviation is 551.0850260383576 
__________________________________________________

## Second Question 

In [59]:
concreteXnorm = concreteX.copy()
for col in concreteXnorm.columns:
    concreteXnorm[f'{col}'] = (concreteXnorm[f'{col}'] - mean(concreteXnorm[f'{col}']))/ stdev(concreteXnorm[f'{col}']) 

In [60]:
mselist2 = []
for count in np.arange(0,50):
    X_train, X_test, y_train, y_split = tts(concreteXnorm, concreteY, test_size=0.3)
    model = regression_model()
    model.fit(X_train, y_train, epochs=50, verbose=0)
    yhat = model.predict(X_test)
    mse = mean_squared_error(yhat,y_split)
    mselist2.append(mse)
    print(count)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49


In [61]:
print('_'*50)
print('The mean of the MSE is {}'.format(mean(mselist2)))
print('_'*50)
print('The Standard Deviation is {} '.format(stdev(mselist2)))
print('_'*50)

__________________________________________________
The mean of the MSE is 369.2589675233575
__________________________________________________
The Standard Deviation is 85.83880699350458 
__________________________________________________


In [79]:
print(f' it has {round(100 - stdev(mselist2)/stdev(mselist)*100,2)}% of imporvment in the std deviation')

 it has 84.42% of imporvment in the std deviation


## Answer 

__________________________________________________
The mean of the MSE is 369.2589675233575
__________________________________________________
The Standard Deviation is 85.83880699350458 
__________________________________________________


 it has 84.42% of imporvment in the std deviation

## Third Question 100 Epoch

In [62]:
mselist3 = []
for count in np.arange(0,50):
    X_train, X_test, y_train, y_split = tts(concreteXnorm, concreteY, test_size=0.3)
    model = regression_model()
    model.fit(X_train, y_train, epochs=100, verbose=0)
    yhat = model.predict(X_test)
    mse = mean_squared_error(yhat,y_split)
    mselist3.append(mse)
    print(count)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49


In [64]:
print('_'*50)
print('The mean of the MSE is {}'.format(mean(mselist3)))
print('_'*50)
print('The Standard Deviation is {} '.format(stdev(mselist3)))
print('_'*50)

__________________________________________________
The mean of the MSE is 168.57864010877597
__________________________________________________
The Standard Deviation is 16.247367249751075 
__________________________________________________


In [80]:
print(f' it has {round(100 - stdev(mselist3)/stdev(mselist2)*100,2)}% of imporvment in the std deviation')

 it has 81.07% of imporvment in the std deviation


## Answer


__________________________________________________
The mean of the MSE is 168.57864010877597
__________________________________________________
The Standard Deviation is 16.247367249751075 
__________________________________________________



 it has 81.07% of imporvment in the std deviation

## Fourth Question 3 Hidden layers

In [65]:
def regression_model4():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(concreteX.shape[1],)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [66]:
mselist4 = []
for count in np.arange(0,50):
    X_train, X_test, y_train, y_split = tts(concreteXnorm, concreteY, test_size=0.3)
    model = regression_model4()
    model.fit(X_train, y_train, epochs=50, verbose=0)
    yhat = model.predict(X_test)
    mse = mean_squared_error(yhat,y_split)
    mselist4.append(mse)
    print(count)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49


In [67]:
print('_'*50)
print('The mean of the MSE is {}'.format(mean(mselist4)))
print('_'*50)
print('The Standard Deviation is {} '.format(stdev(mselist4)))
print('_'*50)

__________________________________________________
The mean of the MSE is 128.4583146175038
__________________________________________________
The Standard Deviation is 14.523611324294263 
__________________________________________________


# answer
__________________________________________________
The mean of the MSE is 128.4583146175038
__________________________________________________
The Standard Deviation is 14.523611324294263 
__________________________________________________

In [81]:
print(f' it has {round(100 - stdev(mselist4)/stdev(mselist3)*100,2)}% of imporvment in the std deviation')

 it has 10.61% of imporvment in the std deviation


# Answer

 it has 10.61% of imporvment in the std deviation